client/src/main/scala/io/delta/sharing/client/auth/CredentialProvider.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.sharing.client.auth
+
+import org.apache.http.HttpHeaders
+import org.apache.http.client.methods.HttpRequestBase
+import org.apache.http.impl.client.CloseableHttpClient
+
+import io.delta.sharing.client.{BearerTokenDeltaSharingProfile, DeltaSharingProfile, OAuthClientCredentialsDeltaSharingProfile}
+
+private[client] object AuthCredentialProviderFactory {
+  def createCredentialProvider(profile: DeltaSharingProfile,
+                               client: CloseableHttpClient): CredentialProvider = {
+    profile match {
+      case oauthProfile: OAuthClientCredentialsDeltaSharingProfile =>
+        OAuthClientCredentialsAuthProvider(client, oauthProfile)
+      case BearerTokenDeltaSharingProfile(_, _, bearerToken, _) =>
+        BearerTokenAuthProvider(bearerToken)
+    }
+  }
+}
+
+private[client] trait CredentialProvider {
+  def addAuthHeader(httpRequest: HttpRequestBase): Unit
+}
+
+private[client] case class BearerTokenAuthProvider(bearerToken: String) extends CredentialProvider {
+  override def addAuthHeader(httpRequest: HttpRequestBase): Unit = {
+    httpRequest.setHeader(HttpHeaders.AUTHORIZATION, s"Bearer $bearerToken")
+  }
+}
+
+private[client] case class
+OAuthClientCredentialsAuthProvider(client: CloseableHttpClient,
+                                   profile: OAuthClientCredentialsDeltaSharingProfile)
+  extends CredentialProvider {
+
+  private[auth] lazy val oauthClient = new OAuthClient(client,
+    profile.tokenEndpoint, profile.clientId, profile.clientSecret, profile.scope)
+
+  private var currentToken: Option[OAuthClientCredentials] = None
+
+  override def addAuthHeader(httpRequest: HttpRequestBase): Unit = {
+    if (currentToken.isEmpty || needsRefresh(currentToken.get)) {
```

Comment:
Nit: refreshToken already checks for this?

Reply:
This is intentional to account for multi-threading before acquiring the synchronization lock. This ensures that if another thread refreshes the token before this thread acquires the synchronization lock, this thread will reuse the refreshed token.